### PR TITLE
[BugFix] forbid rewrite projection with one dict col mapping to multiple output col

### PR DESCRIPTION
**Pull request by packy92**
Fixes #issue
The root cause is missing projection mapping when rewriting projection if there exists one dict col mapping to multiple output col. The  `newProjectMap.put(dictColumn, dictColumn);` only put the only one dict <-> dict col in the map when the old projection contains `col1 <-> dict`, `col2 <-> dict`, `col3 <-> dict`. 

```
if (valueOperator instanceof ColumnRefOperator) {
  ColumnRefOperator stringColumn = (ColumnRefOperator) valueOperator;
  Integer columnId = context.stringColumnIdToDictColumnIds.get(stringColumn.getId());
  ColumnRefOperator dictColumn = context.columnRefFactory.getColumnRef(columnId);

  newProjectMap.put(dictColumn, dictColumn);
  newProjectMap.remove(keyColumn);

  newStringToDicts.put(keyColumn.getId(), dictColumn.getId());
  return;
}
```
When we decode this projection, the decodeOperator can only put one `dict -> col1` entry in the dictToStrings map. So the above operator cannot obtain `col2, col3` input caused NPE when building expr.
```
Map<ColumnRefOperator, ColumnRefOperator> dictToStrings = Maps.newHashMap(); 
for (Map.Entry<Integer, Integer> entry : context.stringColumnIdToDictColumnIds.entrySet()) { 
  dictToStrings.put(context.columnRefFactory.getColumnRef(entry.getValue()), 
  context.columnRefFactory.getColumnRef(entry.getKey())); 
}
```

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.0
  - [x] 2.5
  - [x] 2.4
  - [x] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 10 / 10 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|10|10|100.00%|[]|

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/LowCardinalityTest.java [hunk 1]
```diff
@@ -120,6 +120,65 @@ public static void beforeClass() throws Exception {
                 "\"replication_num\" = \"1\",\n" +
                 "\"in_memory\" = \"false\"\n" +
                 ");");
+        starRocksAssert.withTable("CREATE TABLE `low_card_t1` (\n" +
```
Comment (stdpain) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/LowCardinalityTest.java:123]: could we simplify the case?

Comment (packy92) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/LowCardinalityTest.java:123]: I try use the exist table to rewrite the case, but failed....

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 10 / 10 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|10|10|100.00%|[]|

**Review by stdpain — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by packy92** [comment]
@Mergifyio backport branch-2.5 branch-2.4 branch-2.3

**Pull request merged**